File: src/util/diagnostic.ts
import type { DiagnosticKind } from "../analyzer/diagnostic-kind.js";
import type { Diagnostic, Span } from "../ir/model.js";

export class DiagnosticCollector {
  private readonly items: /* mutable */ Diagnostic[] = [];

  add(kind: DiagnosticKind, message: string, span: Span): void {
    this.items.push({ kind, message, span });
  }

  list(): readonly Diagnostic[] {
    return this.items;
  }
}

File: crates/unsnarl-ir/src/diagnostic.rs
//! Diagnostic record carried in `SerializedIR.diagnostics`.
//!
//! Flattened into a single file (rather than `diagnostic/diagnostic.rs`)
//! to avoid Rust's `module_inception` shape; the topic doesn't fan out
//! into multiple files yet.

use serde::Serialize;

use crate::diagnostic_kind::DiagnosticKind;
use crate::primitive::Span;

#[derive(Serialize)]
pub struct Diagnostic {
    pub kind: DiagnosticKind,
    pub message: String,
    pub span: Span,
}
